### PR TITLE
[FEATURE][text renderer] Support vertical text orientation

### PR DESCRIPTION
**Pull request by nirvn**
## Description
This PR unlocks vertically oriented labels, which should please a large part of the world (Chinese, Japanese, and Korean writers come to mind among others). A little GIF to please the eyes:
peekQZ105Z

The vertical orientation support was added to our QgsTextRenderer class, meaning that any part of QGIS which relies on that to render text will gain vertical orientation support.

On the labelling front, buffers/shadow/background all work, and the verically orientated labels play well with @nyalldawson 's callouts work:
image

I want to get this reviewed now prior to adding tests. 

@andreasneumann , this is all your fault as this was directly triggered by your email on Chinese users :wink: 

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [x] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [x] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit

### CODE REVIEW

**Comment by andreasneumann** [comment]
> @andreasneumann , this is all your fault as this was directly triggered by your email on Chinese users

;-) good that I triggered something. This new feature text-handling be popular in the east.

**Comment by nirvn** [comment]
Here's another example, with label backgrounds:
image

File: src/core/qgstextrenderer.cpp [hunk 1]
```diff
@@ -3217,175 +3282,332 @@ void QgsTextRenderer::drawTextInternal( TextPart drawType,
...
+            context.painter()->setRenderHint( QPainter::TextAntialiasing );
+            context.painter()->drawText( 0, 0, subComponent.text );
+          }
+        }
+      }
+      context.painter()->restore();
+      i++;
     }
+  }
+  else if ( format.orientation() == QgsTextFormat::VerticalOrientation )
```
Comment (nyalldawson) [inline src/core/qgstextrenderer.cpp]: Lots of duplicate code here. Is there anyway to reduce this?

File: src/gui/qgstextformatwidget.cpp [hunk 2]
```diff
@@ -200,6 +200,10 @@ void QgsTextFormatWidget::initWidget()
   connect( chkLineBelow, &QAbstractButton::toggled, this, &QgsTextFormatWidget::updateLinePlacementOptions );
   connect( chkLineOn, &QAbstractButton::toggled, this, &QgsTextFormatWidget::updateLinePlacementOptions );
 
+  mTextOrientationComboBox->addItem( tr( "Horizontal" ), QgsTextFormat::HorizontalOrientation );
```
Comment (nyalldawson) [inline src/gui/qgstextformatwidget.cpp]: Could be nice to have icons for these

File: src/gui/qgstextformatwidget.cpp [hunk 3]
```diff
@@ -200,6 +200,10 @@ void QgsTextFormatWidget::initWidget()
...
+  mTextOrientationComboBox->addItem( tr( "Vertical" ), QgsTextFormat::VerticalOrientation );
+  //connect( mTextOrientationComboBox, static_cast<void ( QComboBox::* )( int )>( &QComboBox::currentIndexChanged ), this, &QgsTextFormatWidget::mTextOrientationComboBox_currentIndexChanged );
```
Comment (nyalldawson) [inline src/gui/qgstextformatwidget.cpp]: Needed?

**Review by nyalldawson — COMMENTED** [review]
Looking good!

File: src/core/qgspallabeling.cpp [hunk 4]
```diff
@@ -1512,16 +1526,32 @@ void QgsPalLayerSettings::calculateLabelSize( const QFontMetricsF *fm, const QSt
...
 
-  for ( const QString &line : multiLineSplit )
-  {
-    w = std::max( w, fm->width( line ) );
+      int maxLineLength = 0;
+      for ( const auto &line : multiLineSplit )
+      {
+        maxLineLength = std::max( maxLineLength, line.length() );
+      }
+      h = labelHeight * maxLineLength + ( maxLineLength - 1 ) * letterSpacing;
```
[the remark this answers is not in the document]
Comment (nirvn) [reply src/core/qgspallabeling.cpp]: @nyalldawson , word spacing has no meaning in the vertically oriented world :)

File: src/core/qgstextrenderer.cpp [hunk 5]
```diff
@@ -1880,6 +1891,7 @@ void QgsTextFormat::readXml( const QDomElement &elem, const QgsReadWriteContext
   {
     d->opacity = ( textStyleElem.attribute( QStringLiteral( "textOpacity" ) ).toDouble() );
   }
+  d->orientation = QgsTextRendererUtils::decodeTextOrientation( textStyleElem.attribute( QStringLiteral( "textOrientation" ) ) );
```
[the remark this answers is not in the document]
Comment (nirvn) [reply src/core/qgstextrenderer.cpp]: @nyalldawson , good to know, should I remove it altogether then?

File: src/gui/qgstextformatwidget.cpp (hunk 2 above)
Comment (nirvn) [reply src/gui/qgstextformatwidget.cpp]: Yes agreed, that's in the to-do list but I'd like to approach this through a series or PRs. It'll hopefully come attached to the next PR.

File: src/gui/qgstextformatwidget.cpp (hunk 3 above)
Comment (nirvn) [reply src/gui/qgstextformatwidget.cpp]: But of course! :wink:

File: src/core/qgstextrenderer.cpp (hunk 1 above)
Comment (nirvn) [reply src/core/qgstextrenderer.cpp]: Yeah, I thought so too, but I can't really see how to reduplicate this. Open to ideas.

For the record: there will be a 3rd vertical orientation mode added in a follow up PR that'll set the vertical orientation based on the rotation. So horizontal, vertical, rotation-dependent / based on rotation (not settled on name yet).

File: src/core/qgstextrenderer.cpp [hunk 6]
```diff
@@ -2559,7 +2579,24 @@ void QgsTextRenderer::drawBuffer( QgsRenderContext &context, const QgsTextRender
...
+    case QgsTextFormat::VerticalOrientation:
+      double letterSpacing = format.scaledFont( context ).letterSpacing();
+      double labelWidth = fontMetrics->maxWidth();
+      const QStringList parts = QgsPalLabeling::splitToGraphemes( component.text );
+      double partYOffset = 0.0;
+      for ( const auto &part : parts )
+      {
+        double partXOffset = ( labelWidth - ( fontMetrics->width( part ) - letterSpacing ) ) / 2;
+        path.addText( partXOffset, partYOffset, format.scaledFont( context ), part );
+        partYOffset += fontMetrics->descent() + fontMetrics->ascent() + letterSpacing;
```
Comment (nyalldawson) [inline src/core/qgstextrenderer.cpp]: Just wondering about this. I realise it's correct logic for scripts with equal character heights, but for others (eg Latin) shouldn't we consider the actual character heights?

Comment (nirvn) [reply src/core/qgstextrenderer.cpp]: I'd rather follow the decisions by much wiser people who codified the behavior in firefox, chrome, et cie and stick to fixed height. 

For one, we need a perfect alignment of characters across each lines all eastern scripts. But its also the case that latin characters would be far less readable with variable heights. Dots would become a tiny... dot and would merge with other characters etc.

**Review by nyalldawson — COMMENTED** [review]
Mind also adding some tests for QgsTextRenderer alone, outside of labeling? This would test the other modes text renderer works in (i.e. rect and point mode)

**Pull request merged**